torch/onnx/_internal/exporter.py

```diff
@@ -382,18 +442,73 @@ def save(
         self,
         destination: Union[str, io.BufferedIOBase],
         *,
+        model_state_dict: Optional[Any] = None,
         serializer: Optional[ExportOutputSerializer] = None,
     ) -> None:
         """Saves the in-memory ONNX model to ``destination`` using specified ``serializer``.
-        If no ``serializer`` is specified, the model will be serialized as Protobuf."""
+
+        Args:
+            destination: The destination to save the ONNX model. It can be either a string or a file-like object.
+                When used with ``model_state_dict``, it must be a string with a full path to the destination.
+                In that case, besides saving the ONNX model, a folder with "_initializers" suffix (without extension)
+                will be created to store the each initializer of the ONNX model in a separate file. For example, if the
+                destination is "/path/model.onnx", the initializers will be saved in "/path/model_initializers/" folder.
+            model_state_dict: The state_dict of the PyTorch model with all weights on it (e.g. `model.state_dict()`).
+                Required when ``enable_fake_mode`` is used but real initializers are needed on the ONNX graph.
+                It can be either a string with the path to a checkpoint or a dictionary with the actual model state.
+            serializer: The serializer to use. If not specified, the model will be serialized as Protobuf.
+        """
 
         if serializer is None:
             serializer = ProtobufExportOutputSerializer()
-        if isinstance(destination, str):
-            with open(destination, "wb") as f:
-                serializer.serialize(self, f)
+
+        # Add initializers when symbolic tracing is enabled
+        _model_state_dict_files = None
+        if model_state_dict is not None:
+            if isinstance(model_state_dict, dict):
+                model_state_dict_file = io.BytesIO()
+                torch.save(model_state_dict, model_state_dict_file)
```

Comment:
Is this method performant? `torch.save`, `torch.load`, `save_model_with_external_data` are being called which feels like more than 1 round trip of read/write occurred.

Reply:
It looks like all the save and loads are to accommodate the API format of previous existing method `save_model_with_external_data`, which expects source model weight/buffers in file paths format. We should clean that up.

Reply:
The read/write does not persist in (slow) disk, but in-memory only, so it should be fine for a export-once-run-many-times approach.

The "save to file" approach with each initializer in a separate file is needed for the really large models which exceeds 2GB. TNLGv4 model, for instance, had this one initializer that was like 1.88GB and by grouping it with other initializers exceeded the 2GB limit.

https://github.com/pytorch/pytorch/pull/103865/files#diff-0795f54fd1f38cfbf2c4a863a4efc9f40f2ea020a2b1612605c361b8d8d35862R524 has a TODO in which I plan to come back to this and improve this part

Reply:
Yup, this is just the start of a series of changes with the goal of exporting really large models. I will create separate tasks to track the next steps